ocs/rook.py

```diff
@@ -0,0 +1,142 @@
+"""
+A module for all rook functionalities and abstractions.
+
+This module has rook related classes, support for functionalities to work with
+rook cluster. This works with assumptions that an OCP cluster is already
+functional and proper configurations are made for interaction.
+"""
+
+import logging
+
+import oc.openshift_ops as ac
+from ocs.pod import Pod
+
+logger = logging.getLogger(__name__)
+
+
+class Rook(object):
+    """
+    High level rook abstraction. This class should handle operator
+    and cluster objects.
+
+    As of now deligated rook environment to corresponding cluster object
+    but in future rookenv should be handled here.Also few operator related ops
+    will be included in this class. Keeping this class lightweight for the first
+    cut.
+
+    Attrs:
+        self._cluster (RookCluster): current cluster object
+    """
+
+    def __init__(self, **config):
+        self._cluster = RookCluster(**config)
+
+    @property
+    def cluster(self):
+        return self._cluster
+
+
+class RookCluster(object):
+    """
+    Handles all cluster related operations from ceph perspective
+
+    This class has depiction of ceph cluster. Contains references to
+    pod objects which represents ceph cluster entities.
+
+    Attributes:
+        _ocs_pods (list) : A list of  ceph cluster related pods
+
+        _cluster_name (str): Name of ceph cluster.
+
+        _namespace (str): openshift Namespace where this cluster lives.
+
+        _api_client (APIClient): api-client used for interacting with openshift.
+
+        _url_prefix (str): url prefix for REST api interaction
+
+    Methods:
+       add_mon(**kw) : add a new mon
+       remove_mon(id): Remove a mon with id id
+       add_osd(**kw):  add a new osd to cluster
+    """
+
+    def __init__(self, **config):
+        # cluster_name is name of cluster in rook of type CephCluster
+        self._cluster_name = config.get('cluster_name', 'rook-ceph')
+        self._namespace = config.get('namespace', 'rook-ceph')
+        # Keeping api_client_name for upcoming PR
+        self._api_client_name = config.get('api_client_name', 'OCRESTClient')
+        self.rook_crd_ver = config.get('rook_crd_ver', 'v1')
+        self._api_client = None         # APIClient object
+        self._url_prefix = config.get(
+            'url_prefix',
+            f'/apis/ceph.rook.io/{self.rook_crd_ver}'
+            f'/namespaces/{self._namespace}/'
+        )
+        self._ocs_pods = list()
+        self._api_client = ac.OCP()  # TODO: APIClient abstractions
+        self.ocs_pod_init()
+
+    @property
+    def cluster_name(self):
+        return self._cluster_name
+
+    @property
+    def namespace(self):
+        return self._namespace
+
+    @property
+    def pods(self):
+        # TODO: Decide whether to return list or yield
```

Comment:
Not sure what advantage we will get here when using generator if we already have the list defined.

Reply:
There is advantage when its a generator, each function call can process one required 'role' and yield that value instead of complete list, if someone wants full list from generator they can always call list()

Reply:
As of now yield looks overkill but considering atleast 20+ pods in ceph cluster (going fwd cluster may grow) and mostly users will be iterating over pods for one or the other operation I made it as generator. If no objection then I will leave as is.

Reply:
I assume we are ok with this.